File: imageselector/src/main/java/com/jhj/imageselector/ImageSelector.java
package com.jhj.imageselector;

import android.app.Activity;
import android.content.Intent;
import android.os.Parcelable;

import com.jhj.imageselector.activityresult.ActivityResult;
import com.jhj.imageselector.bean.LocalMedia;
import com.jhj.imageselector.config.ImageExtra;
import com.jhj.imageselector.ui.ImagePreviewActivity;
import com.jhj.imageselector.ui.ImageSelectorActivity;
import com.jhj.imageselector.utils.ExtensionKt;

import java.util.List;

public class ImageSelector {

    private final Activity mActivity;

    private ImageSelector(Activity activity) {
        this.mActivity = activity;
    }

    public static ImageSelector with(Activity activity) {
        return new ImageSelector(activity);
    }

    public void imageSelected(List<LocalMedia> a, final OnImageSelectedListener listener) {
        Intent intent = new Intent(mActivity, ImageSelectorActivity.class);
        intent.putExtra(ImageExtra.EXTRA_SELECTED_MODE, ImageExtra.MULTI);
        intent.putExtra(ImageExtra.EXTRA_SELECTED_MAX_NUM, 9);
        intent.putExtra(ImageExtra.EXTRA_SELECTED_MIN_NUM, 1);
        ActivityResult.with(mActivity)
                .targentIntent(intent)
                .onResult(new ActivityResult.OnActivityResultListener() {
                    @Override
                    public void onResult(Intent data) {
                        if (data != null) {
                            Object ob = data.getSerializableExtra(ImageExtra.EXTRA_SELECTED_RESULT);
                            if (ob != null) {
                                listener.onSelected((List<LocalMedia>) ob);
                            }
                        }
                    }
                });

      /*  ActivityResult.with(mActivity)
                .targetActivity(ImageSelectorActivity.class)
                .putInt(ImageExtra.EXTRA_SELECTED_MODE, ImageExtra.MULTI)
                .putInt(ImageExtra.EXTRA_SELECTED_MAX_NUM, 9)
                .putInt(ImageExtra.EXTRA_SELECTED_MIN_NUM, 1)
                .putSerializable("a", ExtensionKt.toArrayList(a))
                .onResult(new ActivityResult.OnActivityResultListener() {
                    @Override
                    public void onResult(Intent data) {
                        if (data!= null){
                            Object ob = data.getSerializableExtra(ImageExtra.EXTRA_SELECTED_RESULT);
                            if (ob != null) {
                                listener.onSelected((List<LocalMedia>) ob);
                            }
                        }
                    }
                });*/
        mActivity.overridePendingTransition(R.anim.activity_fade_out, 0);
    }


    public void imageSelected(int selectedMode, int selectedMaxNum, int selectedMinNum, final OnImageSelectedListener listener) {
        ActivityResult.with(mActivity)
                .targetActivity(ImageSelectorActivity.class)
                .putInt(ImageExtra.EXTRA_SELECTED_MODE, selectedMode)
                .putInt(ImageExtra.EXTRA_SELECTED_MAX_NUM, selectedMaxNum)
                .putInt(ImageExtra.EXTRA_SELECTED_MIN_NUM, selectedMinNum)
                .onResult(new ActivityResult.OnActivityResultListener() {
                    @Override
                    public void onResult(Intent data) {
                        List<LocalMedia> localMediaList = (List<LocalMedia>) data.getSerializableExtra(ImageExtra.EXTRA_SELECTED_RESULT);
                        listener.onSelected(localMediaList);
                    }
                });
    }


    public void imagePreview(List<LocalMedia> imageList) {
        imagePreview(imageList, 0);
    }

    public void imagePreview(List<LocalMedia> imageList, int currentIndex) {
        imagePreview(imageList, currentIndex, false);
    }

    public void imagePreview(List<LocalMedia> imageList, int currentIndex, boolean isDelete) {
        Intent intent = new Intent(mActivity, ImagePreviewActivity.class);
        intent.putParcelableArrayListExtra(ImageExtra.IMAGE_LIST, ExtensionKt.<Parcelable>toArrayList(imageList));
        intent.putExtra(ImageExtra.IMAGE_INDEX, currentIndex);
        intent.putExtra(ImageExtra.IMAGE_IS_DELETE, isDelete);
        mActivity.startActivity(intent);
        mActivity.overridePendingTransition(R.anim.activity_fade_out, 0);
    }

    public interface OnImageSelectedListener {
        void onSelected(List<LocalMedia> list);
    }
}

File: imageselector/src/main/java/com/jhj/imageselector/ImageSelector.kt
package com.jhj.imageselector

import android.app.Activity
import com.google.gson.Gson
import com.jhj.imageselector.activityresult.ActivityResult
import com.jhj.imageselector.bean.LocalMedia
import com.jhj.imageselector.config.ImageExtra
import com.jhj.imageselector.ui.ImagePreviewActivity
import com.jhj.imageselector.ui.ImageSelectorActivity
import com.jhj.imageselector.utils.LiveDataBus
import com.jhj.imageselector.utils.toArrayList

object ImageSelector {
    fun camera(mActivity: Activity, body: (List<LocalMedia>) -> Unit) {
        ActivityResult.with(mActivity)
                .targetActivity(ImageSelectorActivity::class.java)
                .putBoolean(ImageExtra.EXTRA_IMAGE_IS_ONLY_CAMERA, true)
                .onResult { data ->
                    val list = data?.getParcelableArrayListExtra<LocalMedia>(ImageExtra.EXTRA_SELECTED_RESULT).orEmpty()
                    body(list)
                }
        mActivity.overridePendingTransition(R.anim.activity_fade_in, R.anim.activity_fade_out)
    }

    @JvmOverloads
    fun singleSelected(mActivity: Activity, localMedia: LocalMedia?, isImageEditable: Boolean = false, body: (List<LocalMedia>) -> Unit) {
        val list = if (localMedia == null) arrayListOf() else arrayListOf(localMedia)
        selected(mActivity, ImageExtra.SINGLE, list, isImageEditable = isImageEditable, body = body)
    }

    @JvmOverloads
    fun multiSelected(mActivity: Activity, imageSelectedList: List<LocalMedia> = arrayListOf(), selectedMaxNum: Int = 9, selectedMinNum: Int = 1, isImageEditable: Boolean = false, body: (List<LocalMedia>) -> Unit) {
        selected(
                mActivity = mActivity,
                imageSelectedList = imageSelectedList,
                selectedMaxNum = selectedMaxNum,
                selectedMinNum = selectedMinNum,
                isImageEditable = isImageEditable,
                body = body)
    }

    @JvmOverloads
    fun selected(mActivity: Activity, selectedMode: Int = ImageExtra.MULTI, imageSelectedList: List<LocalMedia> = arrayListOf(), selectedMaxNum: Int = 9,
                 selectedMinNum: Int = 1, isImageEditable: Boolean = false, body: (List<LocalMedia>) -> Unit = {}) {
        ActivityResult.with(mActivity)
                .targetActivity(ImageSelectorActivity::class.java)
                .putParcelableArrayList(ImageExtra.EXTRA_IMAGE_SELECTED_LIST, imageSelectedList.toArrayList())
                .putInt(ImageExtra.EXTRA_SELECTED_MODE, selectedMode)
                .putInt(ImageExtra.EXTRA_SELECTED_MAX_NUM, selectedMaxNum)
                .putInt(ImageExtra.EXTRA_SELECTED_MIN_NUM, selectedMinNum)
                .onResult { data ->
                    val list = data?.getParcelableArrayListExtra<LocalMedia>(ImageExtra.EXTRA_SELECTED_RESULT).orEmpty()
                    body(list)
                }
        mActivity.overridePendingTransition(R.anim.activity_fade_in, R.anim.activity_fade_out)
    }


    fun preview(mActivity: Activity, imageList: List<LocalMedia>, currentIndex: Int = 0) {
        preview(mActivity, imageList, currentIndex, false, {})
    }


    fun deletePreview(mActivity: Activity, imageList: List<LocalMedia>, currentIndex: Int = 0, body: (List<LocalMedia>) -> Unit) {
        preview(mActivity, imageList, currentIndex, true, body)
    }

    fun preview(mActivity: Activity, imageList: List<LocalMedia>, currentIndex: Int = 0, isDelete: Boolean = false,
                body: (List<LocalMedia>) -> Unit = {}) {
        LiveDataBus.get().with(ImageExtra.EXTRA_IMAGE_LIST).value = Gson().toJson(imageList)
        ActivityResult.with(mActivity)
                .putInt(ImageExtra.EXTRA_IMAGE_INDEX, currentIndex)
                .putBoolean(ImageExtra.EXTRA_IMAGE_IS_DELETE, isDelete)
                .targetActivity(ImagePreviewActivity::class.java)
                .onResult { data ->
                    val list = data?.getParcelableArrayListExtra<LocalMedia>(ImageExtra.EXTRA_SELECTED_RESULT).orEmpty()
                    body(list)
                }

        mActivity.overridePendingTransition(R.anim.activity_fade_in, R.anim.activity_fade_out)
    }


}